meson.setupHook: prefer meson commands over ninja

alyssais opened the pull request:
###### Description of changes

Meson now comes with its own set of commands for building, testing, installing etc., that by default wrap around Ninja.  The reason to prefer using the Meson commands is that they take additional options (e.g. setting custom timeouts for tests — my motivation for this change).

Here, I've modified the Meson setup hook so instead of Ninja being run automatically, Meson's equivalent commands will be if Meson is present.  This only happens when Meson is used to configure, to avoid starting to run Meson directly when dealing with custom build systems that wrap around Meson, like QEMU's.

Naturally the Meson commands don't support entirely the same set of options that the Ninja ones did, but I checked through Nixpkgs to find any packages using Meson that used any options that wouldn't be picked up by this new system.  I only found one, and it was just setting `checkTarget = "test"`, which is the default value for Ninja and has no Meson equivalent (because we directly tell Meson to run the tests rather than going through a generic job system like Ninja).

I've tested building all the way to qemu_kvm, which I chose because it has lots of dependencies built with Meson, and also has its own custom Meson-wrapping build system so I could be sure that also still worked.

Fixes, partially: https://github.com/NixOS/nixpkgs/issues/113829

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

jtojnar commented on pkgs/development/tools/build-managers/meson/setup-hook.sh:
```diff
@@ -31,7 +31,69 @@ mesonConfigurePhase() {
+
+    localFlagsArray=(
+        -j$NIX_BUILD_CORES
+        $mesonBuildFlags "${mesonBuildFlags[@]}"
```
Won’t this be partly duplicated?

jtojnar commented on pkgs/development/tools/build-managers/meson/setup-hook.sh:
```diff
@@ -31,7 +31,69 @@ mesonConfigurePhase() {
+    )
+
+    echoCmd 'build flags' "${flagsArray[@]}"
+    TERM=dumb meson compile "${flagsArray[@]}"
```
Why the `TERM`? `setup` does not set this.

alyssais replied:
It's for Ninja (which `setup` doesn't run). https://github.com/NixOS/nixpkgs/pull/179027

jtojnar replied:
Might be nice to add a comment, or at least mention it in the commit message then.

eli-schwartz replied:
It's only needed for ninja, which could just inherit the default ninja handling.

Why set it for test/install? Those don't generally build anything anyway (it's already built when you called ninja) and TERM=dumb doesn't seem to be targeted at meson's test harness.

alyssais replied:
I wasn't sure if test/install still ended up calling Ninja.  I expected at least install to.  If they don't I can remove it.

jtojnar commented on pkgs/development/tools/build-managers/meson/setup-hook.sh:
```diff
@@ -31,7 +31,69 @@ mesonConfigurePhase() {
     configurePhase=mesonConfigurePhase
 fi
+
+if [ -z "${dontUseMesonBuild-${dontUseMesonConfigure-}}" -a -z "${buildPhase-}" ]; then
```
What if ninja setup hook runs first?

alyssais replied:
That's a very good point.  When I was originally thinking about this, I assumed that the `dontUseNinjaBuild` etc. checks wouldn't happen until their respective phases, so setting them immediately in the meson hook would always run first, regardless of ordering.  But that's not actually how it works, so I'm not sure what to do here.  Any ideas?

alyssais replied:
I suppose one thing we could do would be to, in `mesonConfigurePhase`, do something like

```bash
if [[ $checkPhase = ninjaCheckPhase && -z $dontUseMesonCheck ]]; then
    checkPhase=mesonCheckPhase
fi
```

It doesn't feel as nice, but it would get around the ordering problem, and it wouldn't make it any more difficult to force the ninja version if you needed to because you'd still just `dontUseMesonCheck = true`.

jtojnar replied:
Yeah, I do not see many other options. It is still better than [what we do in g-i setup hook](https://github.com/NixOS/nixpkgs/blob/d19f38fda982d507684b65efcb00e4b143597843/pkgs/development/libraries/gobject-introspection/setup-hook.sh#L21-L28).

jtojnar replied:
Possibly, we could use an auxiliary variable `usingImplicitNinjaCheckPhase=1` added to ninja setup hook to distinguish it from user explicitly setting `checkPhase=ninjaCheckPhase` but I probably would not bother.

jtojnar reviewed (commented):
Great, wanted this for a while now: https://github.com/NixOS/nixpkgs/issues/113829

eli-schwartz commented on pkgs/development/tools/build-managers/meson/setup-hook.sh:
```diff
@@ -31,7 +31,69 @@ mesonConfigurePhase() {
     runHook postConfigure
 }
 
+mesonBuildPhase() {
```
I would skip this one, you don't need it.

alyssais replied:
Yeah I mostly added this because I was worried it would be more confusing if meson was called directly for some tasks but not others.  I'd be happy to drop it too.  Whatever @jtojnar would prefer as the package maintainer.

eli-schwartz replied:
I like the "less code is better" approach :) and it means you don't have to sync ninja handling over from elsewhere...

eli-schwartz reviewed (changes requested):
> The reason to prefer using the Meson commands is that they take additional options (e.g. setting custom timeouts for tests — my motivation for this change).

- The functionality that is implemented in python (`meson foobar` subcommands) come with convenience targets in ninja, that run the python tools.
- The functionality that is implemented in ninja comes with a convenience target in python, that simply runs ninja.

Practically speaking, this means that if you want the default functionality, you can more or less use either entry point interchangeably.

The difference between the two is that building is purely ninja, and everything else is implemented in python. So it makes sense to run `meson test` and `meson install` directly if you want to pass additional options, but there's no advantage to running `meson compile <args>` since it will do nothing other than subprocess out to `ninja <args>`.

... Oh wait, actually that's not quite true. On Windows, `meson compile` will first run vcvarsall.bat to set up an MSVC dev cmd environment, if you're using the MSVC compiler and didn't start off in one. This functionality has no reason to exist except that ninja's feature request to not be broken by default, but rather to include functionality for loading an environment block at startup, is in limbo. You don't need this if you use a sane operating system, or if you use a mingw-based compiler even on Windows.

alyssais commented:
@eli-schwartz thanks for the clarification/explanation! I appreciate it.

alyssais commented on pkgs/development/tools/build-managers/meson/setup-hook.sh:
```diff
@@ -31,7 +31,69 @@ mesonConfigurePhase() {
+mesonBuildPhase() {
+    runHook preBuild
+
+    localFlagsArray=(
```
Yes :)

alyssais commented:
v2:

- Drop `mesonBuildPhase`.
- Move phase overrides into `mesonConfigurePhase` to avoid setup hook ordering problems.
- Don't set `TERM=dumb` for check/install (which ends up being not at all since we dropped `mesonBuildPhase`).
- Fix the flags typos noticed by @jtojnar.

I'll do another test build through qemu_kvm, but once review has settled down because it takes a long time.

alyssais commented:
v3:

- Don't try to pass -j$NIX_BUILD_CORES to meson test, as it does not support it.

- Pass --no-rebuild to meson test and meson install, to avoid rerunning Ninja, which will never have any work to do.

alyssais commented:
Tested building through to `qemu_kvm` again.

jtojnar commented on pkgs/development/tools/build-managers/meson/setup-hook.sh:
```diff
@@ -28,9 +28,39 @@ mesonConfigurePhase() {
         echo "meson: enabled parallel building"
     fi
 
+    if [[ ${checkPhase-ninjaCheckPhase} = ninjaCheckPhase && -z $dontUseMesonCheck ]]; then
+        checkPhase=mesonCheckPhase
+    fi
```
Having this in `configurePhase` will only work when it is not overridden. But I probably would not bother with this too much until someone comes up with an use case.

alyssais replied:
Yeah I think that's fine.

Artturin replied:
I found only one usage of an overridden configurePhase with meson.
https://github.com/NixOS/nixpkgs/pull/259835

jtojnar reviewed (commented):
LGTM, could you please update the docs https://github.com/NixOS/nixpkgs/blob/6de7b51edd2f868dc0eb05724ea148fd0efcb3a0/doc/hooks/meson.section.md and https://github.com/NixOS/nixpkgs/blob/6de7b51edd2f868dc0eb05724ea148fd0efcb3a0/doc/hooks/ninja.section.md

jtojnar commented on doc/hooks/meson.section.md:
```diff
@@ -1,17 +1,25 @@
 ### Meson {#meson}
 
-Overrides the configure phase to run meson to generate Ninja files. To run these files, you should accompany Meson with ninja. By default, `enableParallelBuilding` is enabled as Meson supports parallel building almost everywhere.
+Overrides the configure, check, and install phases to run `meson setup`, `meson test`, and `meson install`. For a meson-compatible build phase, you should accompany Meson with ninja. By default, `enableParallelBuilding` is enabled as Meson supports parallel building almost everywhere.
```
It is not really obvious to me what Meson-compatible build phase is supposed to mean. Maybe something like the following:

```suggestion
Overrides the configure, check, and install phases to run `meson setup`, `meson test`, and `meson install`.

Meson is a meta-build system so you will need a secondary build system to run the generated build files in build phase. In Nixpkgs context, you will want to accompany Meson with ninja, which provides a [setup hook](#ninja) registering a ninja-based build phase.

By default, `enableParallelBuilding` is enabled as Meson supports parallel building almost everywhere.
```

alyssais replied:
Incorporated.

alyssais committed:
meson.setupHook: prefer meson commands over ninja

Meson now comes with its own set of commands for building, testing,
installing etc., that by default wrap around Ninja.  The reason to
prefer using the Meson commands is that they take additional
options (e.g. setting custom timeouts for tests — my motivation for
this change).

Here, I've modified the Meson setup hook so that Meson's test and
install commands will be used instead of Ninja's when Meson's
configurePhase is used.  This restriction is to avoid starting to run
Meson directly when dealing with custom build systems that wrap around
Meson, like QEMU's.  We don't use meson's compile command, as it just
runs ninja, and that's handled fine by the existing Ninja setup hook.

Naturally the Meson commands don't support entirely the same set of
options that the Ninja ones did, but I checked through Nixpkgs to find
any packages using Meson that used any options that wouldn't be picked
up by this new system.  I only found one, and it was just setting
checkTarget = "test", which is the default value for Ninja and has no
Meson equivalent (because we directly tell Meson to run the tests
rather than going through a generic job system like Ninja).

Link: https://github.com/NixOS/nixpkgs/issues/113829
Co-authored-by: Jan Tojnar <jtojnar@gmail.com>

alyssais commented:
@OfBorg eval

alyssais commented:
OfBorg failure is just it failing a test in aws-sdk-cpp.

jtojnar reviewed (approved):
Perfect, thanks.
